runtime/gc_include/ObjectAllocationAPI.hpp

```diff
@@ -334,19 +332,25 @@ class MM_ObjectAllocationAPI
 			if (initializeSlots) {
 				memset(objectHeader + 1, 0, dataSize);
 			}
-			if (initializeLockWord) {
-				*(U_32*)J9OBJECT_MONITOR_EA(currentThread, instance) = OBJECT_HEADER_LOCK_RESERVED;
-			}
 		} else {
 			J9ObjectFull *objectHeader = (J9ObjectFull*) instance;
 			objectHeader->clazz = (UDATA)clazz;
 			if (initializeSlots) {
 				memset(objectHeader + 1, 0, dataSize);
 			}
-			if (initializeLockWord) {
-				*(UDATA*)J9OBJECT_MONITOR_EA(currentThread, instance) = OBJECT_HEADER_LOCK_RESERVED;
+		}
+
+		if (initializeSlots) {
+			if (LN_HAS_LOCKWORD(currentThread, instance)) {
+				j9objectmonitor_t initialLockword = VM_ObjectMonitor::getInitialLockword(currentThread->javaVM, clazz);
```

Comment:
Is this code breaks run time Compressed/Full support added to this function already?

Reply:
No, it should be compatible. The initial lockword value can only be one of 3 raw values: 0 (never reserve), 4 (go to Reserved state on first lock) or 8 (go to Learning state on first lock). It never sets the TID field so it works for both 32 bit and 64 bit lockwords.

A few lines later the macro `J9_STORE_LOCKWORD` may be used to set the new lockword. `J9_STORE_LOCKWORD` is defined in `j9accessbarrier.h` and uses the `J9VMTHREAD_COMPRESS_OBJECT_REFERENCES` macro to determine if compressed refs are used or not. The width of the store is adjusted accordingly.

Reply:
ok, I see an implementation of `J9_STORE_LOCKWORD`... Logically looks correct however would introduce check of `if (J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread))` for second time

Reply:
also there is difference `J9_STORE_LOCKWORD` definition uses  extra `volatile` declaration. I am not sure can it have performance impact in this case.

Reply:
Do you think it would be better to change this to a non-volatile store? There's also changes in `objectMonitorExit` in `monhelpers.c` where I changed the code to use `J9_STORE_LOCKWORD` that may also need to be changed to use non-volatile stores.

Reply:
I don't think it is necessary to use volatile store here in sequence of Object initialization, since there will be memory barrier later anyway.